## Data Analysis in Python
_Author: Georgy Tarasenko_

## Seminar 9

```pandas``` is a Python library containing a set of functions and specialized data structures that have been designed to help Python programmers to perform data analysis tasks in a structured way (all documentation can be find [here](https://pandas.pydata.org/)).

Most of the things that `pandas` can do can be obtained with basic Python, but the collected set of pandas functions and data structure makes the data analysis tasks more consistent in terms of syntax and therefore aids readabilty. In addition, `pandas` framework is specifically constructed to handle big data.

As pandas is not a built-in library, we have to install it:

In [ ]:
!pip install pandas

Then, to access its tools we have to import this library, we can use `pd` as a shorther from of `pandas` for the purpose of simplicity

In [1]:
import pandas as pd

#### Pandas series: core of pandas
Basic data type in pandas package is `pd.Series`: *mutable* and *ordered*. You can create `pd.Series` using `pd.Series()` function:

In [ ]:
my_series = pd.Series([5, 2, 9])
print(my_series) # indices and values

In [ ]:
type(my_series)

`pd.Series` is able to store only ONE data type (e. g. only `int`, `float` or `str`). It can be created from any sequence:

In [ ]:
from_list = pd.Series([5, 2, 9])
from_tuple = pd.Series((1, 2, 3))

In [ ]:
print(from_tuple)

In [ ]:
from_list = pd.Series([1, 2, '3'])
print(from_list)

You can manually specify indices for your `pd.Series`:

In [ ]:
pd.Series([5, 2, 9], index=[2, 3, 1]) # not necessarily begin with 0 or even be ordered (mix of list and dict indices)...

In [ ]:
s = pd.Series([5, 2, 9], index=[2, 3, 1])
s[0:2]

In [ ]:
pd.Series([5, 2, 9], index=['a', 'b', 'c']) # ...even strings

In [2]:
from_dict = pd.Series({'a': 1, 'b': 2, 'c': 3}) # keys become indices
print(from_dict)

a    1
b    2
c    3
dtype: int64


**Notice.** `pandas` allows non-unique indices, BE CAREFUL.

In [ ]:
pd.Series([5, 2, 9], index=['a', 'a', 'a'])

In [ ]:
tuple1 = (4, 1, 7)
indices = ['a', 'b', 'c']

pd.Series(tuple1, index=indices)

In [ ]:
check = pd.Series([5, 1, 0, 3, 4], index=[6, 1, 0, 4, 2])

In [ ]:
check[2:4]

#### DataFrame

Basic type of data in pandas is `DataFrame`, a table conisting of rows and columns. Each column is `pd.Series`.

We can create our own dataframe by applying `DataFrame()` function to a dictionary where a key is a column name while value is a sequence of objects. Consequently, the length of each sequence should be the same. 

#### Creating our own dataframe

Let's create a dataframe that contains information on the product reviews for two users. We will apply `DataFrame()` function to a dictionary. We use `index` argument to specify the row names.

In [ ]:
reviews_dict = {'Bob': ['I liked it.', 'It was awful.'], 'Sue': ['Pretty good.', 'Bland.']}

reviews = pd.DataFrame({'Bob': ['I liked it.', 'It was awful.'], 
              'Sue': ['Pretty good.', 'Bland.']},
             index=['Product A', 'Product B'])
reviews

In [3]:
spending_dict = {
    'names': ['Ioann', 'Diana', 'Lisa', 'Nastya', 'Sofia', 'Kate', 'Alex'],
    'age': [22, 20, 19, 43, 78, 26, 50],
    'spendings': [2000, 1500, 1000, 20000, 500, 4000, 15000]
}

spending_df = pd.DataFrame(spending_dict)
spending_df

names  age  spendings
0   Ioann   22       2000
1   Diana   20       1500
2    Lisa   19       1000
3  Nastya   43      20000
4   Sofia   78        500
5    Kate   26       4000
6    Alex   50      15000

Yet, in real life tasks we rather deal with external data that was previously collected by us - *primary data* - or by other people (researchers, institutions) - *secondary data*.

#### Reading data frames

`os` module in Python allows to access information on your computer's operational system (e.g. uploading files):

- method `os.chdir(your_path)` sets your current working directory;
- method `os.getcwd()` displays the working directory that you have for now.

In [ ]:
import os
os.chdir('/Users/georgetarasenko/Desktop')
os.getcwd()

Now let's try to import our first dataframe. One of the most widespread extensions is `.csv` - comma separated value - a file where each table cell value is separated by a symbol such as comma or point with comma or separated via tabulation.

If we want to import `.csv` extension data we apply `pd.read_csv()` method inside which we specify the directory of the file (or just its name is the directory was set previously). Besides that we can specify the separator via `sep` argument, `index_col = 0` allows us to make a first column from a dataset to be row names.

We will working with the latest issue of the [**World Bank statistics**](https://data.worldbank.org/) and [**Polity IV project**](https://www.systemicpeace.org/polity/polity4.htm).

In [5]:
wb_data = pd.read_csv('wb_data.csv', index_col=0)
wb_data

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
1    Afghanistan      AFG          8     519.884766  36296400.0       -1.0   
2        Albania      ALB          1    4531.020996   2873457.0        9.0   
3        Algeria      DZA          3    4111.293945  41389200.0        2.0   
4        Andorra      AND          5   38962.878906     77001.0        NaN   
5         Angola      AGO          4    4095.812988  29816748.0       -2.0   
..           ...      ...        ...            ...         ...        ...   
190   Uzbekistan      UZB          1    1826.566895  32388600.0       -9.0   
191    Venezuela      VEN          2   16054.490234  29390408.0       -3.0   
192        Samoa      WSM          9    4259.765137    195352.0        NaN   
193        Yemen      YEM          3     960.528503  27834820.0        0.0   
194       Zambia      ZMB          4    1534.866699  16853688.0        6.0   

     wdi_oilrent  
1       0.001213  
2       1.367729  
3      12.361850  
4            NaN  
5      15.443859  
..           ...  
190     0.721004  
191     9.832438  
192     0.000000  
193     0.791113  
194     0.000000  

[194 rows x 7 columns]

Let's discuss our data. It is data collected by the World Bank combined with Polity Project for 194 countries in 2021.

We have the following variables (columns):
- `cname` - country name;
- `ccodealp` - country code
- `ht_region` - region
- `wdi_gdpcapcur` - GDP per capita 
- `wdi_pop` - population
- `p_polity2` - Index of democracy by Polity IV Project
- `wdi_oilrent` - Share of Oil rents (% of GDP)

We can check the first `n` rows of a data frame through `head()` method:

In [6]:
type(wb_data)

pandas.core.frame.DataFrame

In [8]:
wb_data.head(9) #first 5 rows are displayed automatically

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  \
1          Afghanistan      AFG          8     519.884766  36296400.0   
2              Albania      ALB          1    4531.020996   2873457.0   
3              Algeria      DZA          3    4111.293945  41389200.0   
4              Andorra      AND          5   38962.878906     77001.0   
5               Angola      AGO          4    4095.812988  29816748.0   
6  Antigua and Barbuda      ATG         10   15383.415039     95426.0   
7           Azerbaijan      AZE          1    4147.089844   9854033.0   
8            Argentina      ARG          2   14613.041992  44044812.0   
9            Australia      AUS          5   54027.968750  24601860.0   

   p_polity2  wdi_oilrent  
1       -1.0     0.001213  
2        9.0     1.367729  
3        2.0    12.361850  
4        NaN          NaN  
5       -2.0    15.443859  
6        NaN     0.000000  
7       -7.0    17.816431  
8        9.0     0.727118  
9       10.0     0.210424

Last `n` rows are displayed as follows:

In [9]:
wb_data.tail()

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
190  Uzbekistan      UZB          1    1826.566895  32388600.0       -9.0   
191   Venezuela      VEN          2   16054.490234  29390408.0       -3.0   
192       Samoa      WSM          9    4259.765137    195352.0        NaN   
193       Yemen      YEM          3     960.528503  27834820.0        0.0   
194      Zambia      ZMB          4    1534.866699  16853688.0        6.0   

     wdi_oilrent  
190     0.721004  
191     9.832438  
192     0.000000  
193     0.791113  
194     0.000000

Other basic commands:

In [10]:
wb_data.info() # method (narrow function, uses parentheses)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 194 entries, 1 to 194
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   cname          194 non-null    object 
 1   ccodealp       194 non-null    object 
 2   ht_region      194 non-null    int64  
 3   wdi_gdpcapcur  189 non-null    float64
 4   wdi_pop        192 non-null    float64
 5   p_polity2      165 non-null    float64
 6   wdi_oilrent    183 non-null    float64
dtypes: float64(4), int64(1), object(2)
memory usage: 12.1+ KB


In [11]:
wb_data.dtypes # attribute (just information, no parentheses)

cname             object
ccodealp          object
ht_region          int64
wdi_gdpcapcur    float64
wdi_pop          float64
p_polity2        float64
wdi_oilrent      float64
dtype: object

In [12]:
wb_data.head()

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
1  Afghanistan      AFG          8     519.884766  36296400.0       -1.0   
2      Albania      ALB          1    4531.020996   2873457.0        9.0   
3      Algeria      DZA          3    4111.293945  41389200.0        2.0   
4      Andorra      AND          5   38962.878906     77001.0        NaN   
5       Angola      AGO          4    4095.812988  29816748.0       -2.0   

   wdi_oilrent  
1     0.001213  
2     1.367729  
3    12.361850  
4          NaN  
5    15.443859

In [13]:
len(wb_data)

194

In [14]:
wb_data.shape

(194, 7)

In [15]:
wb_data.columns

Index(['cname', 'ccodealp', 'ht_region', 'wdi_gdpcapcur', 'wdi_pop',
       'p_polity2', 'wdi_oilrent'],
      dtype='object')

In [ ]:
# How many rows?
print(len(wb_data))
# How many rows and columns - returned as a tuple
print(wb_data.shape)
#How many 'cells' in the table
print(wb_data.size)
# What are the column names
print(wb_data.columns)
# what are the data types of the columns?
print(wb_data.dtypes)

#### Selecting rows and columns in pandas

In [16]:
wb_data.head()

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
1  Afghanistan      AFG          8     519.884766  36296400.0       -1.0   
2      Albania      ALB          1    4531.020996   2873457.0        9.0   
3      Algeria      DZA          3    4111.293945  41389200.0        2.0   
4      Andorra      AND          5   38962.878906     77001.0        NaN   
5       Angola      AGO          4    4095.812988  29816748.0       -2.0   

   wdi_oilrent  
1     0.001213  
2     1.367729  
3    12.361850  
4          NaN  
5    15.443859

In [17]:
wb_data['cname']

1      Afghanistan
2          Albania
3          Algeria
4          Andorra
5           Angola
          ...     
190     Uzbekistan
191      Venezuela
192          Samoa
193          Yemen
194         Zambia
Name: cname, Length: 194, dtype: object

In [ ]:
wb_data[['cname']]

In [18]:
columns_to_choose = ['cname', 'p_polity2', 'wdi_oilrent']

wb_data[columns_to_choose]

cname  p_polity2  wdi_oilrent
1    Afghanistan       -1.0     0.001213
2        Albania        9.0     1.367729
3        Algeria        2.0    12.361850
4        Andorra        NaN          NaN
5         Angola       -2.0    15.443859
..           ...        ...          ...
190   Uzbekistan       -9.0     0.721004
191    Venezuela       -3.0     9.832438
192        Samoa        NaN     0.000000
193        Yemen        0.0     0.791113
194       Zambia        6.0     0.000000

[194 rows x 3 columns]

In [21]:
wb_data[4:9:2]

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
5      Angola      AGO          4    4095.812988  29816748.0       -2.0   
7  Azerbaijan      AZE          1    4147.089844   9854033.0       -7.0   
9   Australia      AUS          5   54027.968750  24601860.0       10.0   

   wdi_oilrent  
5    15.443859  
7    17.816431  
9     0.210424

In [ ]:
wb_data[4:9:2]

In [ ]:
# Both of these statements are the same
print(wb_data['cname'])
# and
print(wb_data.cname)

In [22]:
wb_data.loc[1:6, ['cname', 'ccodealp']]

cname ccodealp
1          Afghanistan      AFG
2              Albania      ALB
3              Algeria      DZA
4              Andorra      AND
5               Angola      AGO
6  Antigua and Barbuda      ATG

In [23]:
wb_data.head()

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
1  Afghanistan      AFG          8     519.884766  36296400.0       -1.0   
2      Albania      ALB          1    4531.020996   2873457.0        9.0   
3      Algeria      DZA          3    4111.293945  41389200.0        2.0   
4      Andorra      AND          5   38962.878906     77001.0        NaN   
5       Angola      AGO          4    4095.812988  29816748.0       -2.0   

   wdi_oilrent  
1     0.001213  
2     1.367729  
3    12.361850  
4          NaN  
5    15.443859

In [24]:
wb_data.loc[1:2, ['ht_region', 'wdi_gdpcapcur']]

ht_region  wdi_gdpcapcur
1          8     519.884766
2          1    4531.020996

In [25]:
wb_data.tail()

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
190  Uzbekistan      UZB          1    1826.566895  32388600.0       -9.0   
191   Venezuela      VEN          2   16054.490234  29390408.0       -3.0   
192       Samoa      WSM          9    4259.765137    195352.0        NaN   
193       Yemen      YEM          3     960.528503  27834820.0        0.0   
194      Zambia      ZMB          4    1534.866699  16853688.0        6.0   

     wdi_oilrent  
190     0.721004  
191     9.832438  
192     0.000000  
193     0.791113  
194     0.000000

In [26]:
wb_data.iloc[-5::1, :2]

cname ccodealp
190  Uzbekistan      UZB
191   Venezuela      VEN
192       Samoa      WSM
193       Yemen      YEM
194      Zambia      ZMB

Selecting multiple columns:

In [ ]:
print(wb_data[['cname', 'ccodealp', 'wdi_pop']])

`iloc` method allows to slice the dataframe:

In [ ]:
wb_data.iloc[0:9, 0:3] #range of rows before comma, range of column after comma

Use `rename` method to rename your columns:

In [ ]:
wb_data.rename(columns={'wdi_pop': 'popul'}) #name changed from 'wdi_pop' to 'popul'

Basics of sorting can be done through `sort_values` method: 

In [ ]:
# Sorting
wb_data.sort_values(by='wdi_gdpcapcur', ascending=False, na_position='last')

Aggregating data by groups:

In [ ]:
wb_data.groupby('ht_region').mean()['p_polity2'].sort_values()

In [ ]:
wb_data_rich_dem = wb_data[(wb_data.p_polity2 >= 4) & (wb_data.wdi_gdpcapcur > 30000)]
wb_data_rich_dem

In [ ]:
#From exponentiated to normal values
pd.set_option('display.float_format', '{:.2f}'.format)

In [27]:
wb_data.head()

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
1  Afghanistan      AFG          8     519.884766  36296400.0       -1.0   
2      Albania      ALB          1    4531.020996   2873457.0        9.0   
3      Algeria      DZA          3    4111.293945  41389200.0        2.0   
4      Andorra      AND          5   38962.878906     77001.0        NaN   
5       Angola      AGO          4    4095.812988  29816748.0       -2.0   

   wdi_oilrent  
1     0.001213  
2     1.367729  
3    12.361850  
4          NaN  
5    15.443859

In [29]:
wb_data.rename(columns={'ccodealp': 'country_code', 'cname': 'country_name'})

country_name country_code  ht_region  wdi_gdpcapcur     wdi_pop  \
1    Afghanistan          AFG          8     519.884766  36296400.0   
2        Albania          ALB          1    4531.020996   2873457.0   
3        Algeria          DZA          3    4111.293945  41389200.0   
4        Andorra          AND          5   38962.878906     77001.0   
5         Angola          AGO          4    4095.812988  29816748.0   
..           ...          ...        ...            ...         ...   
190   Uzbekistan          UZB          1    1826.566895  32388600.0   
191    Venezuela          VEN          2   16054.490234  29390408.0   
192        Samoa          WSM          9    4259.765137    195352.0   
193        Yemen          YEM          3     960.528503  27834820.0   
194       Zambia          ZMB          4    1534.866699  16853688.0   

     p_polity2  wdi_oilrent  
1         -1.0     0.001213  
2          9.0     1.367729  
3          2.0    12.361850  
4          NaN          NaN  
5         -2.0    15.443859  
..         ...          ...  
190       -9.0     0.721004  
191       -3.0     9.832438  
192        NaN     0.000000  
193        0.0     0.791113  
194        6.0     0.000000  

[194 rows x 7 columns]

In [ ]:
wb_data.head()

In [30]:
wb_data.sort_values(by=['wdi_gdpcapcur'])

cname ccodealp  ht_region  wdi_gdpcapcur  \
27                      Burundi      BDI          4     292.997620   
104                      Malawi      MWI          4     356.717560   
33     Central African Republic      CAF          4     450.900238   
117                  Mozambique      MOZ          4     461.415100   
42   Congo, Democratic Republic      COD          4     467.074249   
..                          ...      ...        ...            ...   
38                       Taiwan      TWN          6            NaN   
56                      Eritrea      ERI          4            NaN   
90                 Korea, North      PRK          6            NaN   
160                     Somalia      SOM          4            NaN   
170                       Syria      SYR          3            NaN   

        wdi_pop  p_polity2  wdi_oilrent  
27   10827024.0       -1.0     0.000000  
104  17670260.0        6.0     0.000000  
33    4596028.0        6.0     0.000000  
117  28649008.0        5.0     0.096506  
42   81398768.0       -3.0     0.565558  
..          ...        ...          ...  
38          NaN       10.0          NaN  
56          NaN       -7.0          NaN  
90   25429984.0      -10.0          NaN  
160  14589119.0        5.0          NaN  
170  17068002.0       -9.0          NaN  

[194 rows x 7 columns]

In [31]:
wb_data.sort_values(by=['p_polity2'])

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  \
12                 Bahrain      BHR          3   23742.984375   1494074.0   
151           Saudi Arabia      SAU          3   20803.746094  33099148.0   
142                  Qatar      QAT          3   59124.933594   2724724.0   
90            Korea, North      PRK          6            NaN  25429984.0   
170                  Syria      SYR          3            NaN  17068002.0   
..                     ...      ...        ...            ...         ...   
150  Sao Tome and Principe      STP          4    1813.783203    207089.0   
154             Seychelles      SYC          4   15906.083984     95843.0   
174                  Tonga      TON          9    4513.609375    101998.0   
180                 Tuvalu      TUV          9    3572.493652     11370.0   
192                  Samoa      WSM          9    4259.765137    195352.0   

     p_polity2  wdi_oilrent  
12       -10.0     2.013305  
151      -10.0    23.053974  
142      -10.0    14.241599  
90       -10.0          NaN  
170       -9.0          NaN  
..         ...          ...  
150        NaN     0.000000  
154        NaN     0.000000  
174        NaN     0.000000  
180        NaN     0.000000  
192        NaN     0.000000  

[194 rows x 7 columns]

#### Basic characteristics of your data:

In [32]:
wb_data['wdi_pop']

1      36296400.0
2       2873457.0
3      41389200.0
4         77001.0
5      29816748.0
          ...    
190    32388600.0
191    29390408.0
192      195352.0
193    27834820.0
194    16853688.0
Name: wdi_pop, Length: 194, dtype: float64

In [33]:
wb_data['wdi_pop'].mean()


38868002.166666664

In [ ]:
wb_data['wdi_oilrent'].median()


In [ ]:
wb_data['wdi_gdpcapcur'].min()

In [ ]:
groups.get_group(1)['wdi_gdpcapcur'].min()

In [ ]:
wb_data['ht_region'].value_counts()

### Practice 

We will us the same dataset for the practice

##### 1 Rename your regions using the following structure and information from the data decription. The format of naming is on you.

The structure:

```df['column name'] = df['column name'].replace(['1st old value','2nd old value',...],'new value')```

In [ ]:
#your code here

##### 2 Print countries with the share of oil_rents is higher than 1% of GDP, and Polity Index of Democracy above 1. Then, display these countries only for Latin America

In [ ]:
#your code here

##### 3 Using `get_group()` method calculate the mean gdp per capita in Sub Saharan Africa and Western Europe / North America. Using sorting to identify which country is economically poorest in Latin America.

In [ ]:
#your code here

In [34]:
wb_data.head()

cname ccodealp  ht_region  wdi_gdpcapcur     wdi_pop  p_polity2  \
1  Afghanistan      AFG          8     519.884766  36296400.0       -1.0   
2      Albania      ALB          1    4531.020996   2873457.0        9.0   
3      Algeria      DZA          3    4111.293945  41389200.0        2.0   
4      Andorra      AND          5   38962.878906     77001.0        NaN   
5       Angola      AGO          4    4095.812988  29816748.0       -2.0   

   wdi_oilrent  
1     0.001213  
2     1.367729  
3    12.361850  
4          NaN  
5    15.443859

**Mean value of wdi_gdpcapcur for last 10 countries.**

In [37]:
wb_data['wdi_gdpcapcur'][-10::1].mean()

14412.184014892578

**Find 5 most democratic countries.**

In [42]:
wb_data.sort_values(by=['p_polity2'], ascending=False)[:5]['cname']

169    Switzerland
124    New Zealand
157       Slovakia
31          Canada
32      Cape Verde
Name: cname, dtype: object